libs/dal/src/repositories/base-repository.ts

```diff
@@ -320,6 +325,150 @@ export class BaseRepository<T_DBModel, T_MappedEntity, T_Enforcement> {
   async withTransaction(fn: Parameters<ClientSession['withTransaction']>[0]) {
     return (await this._model.db.startSession()).withTransaction(fn);
   }
+
+  async findWithCursorBasedPagination({
+    query = {} as FilterQuery<T_DBModel> & T_Enforcement,
+    limit,
+    before,
+    after,
+    sortBy,
+    sortDirection = DirectionEnum.DESC,
+    paginateField,
+    enhanceQuery,
+  }: {
+    query?: FilterQuery<T_DBModel> & T_Enforcement;
+    limit: number;
+    before?: { sortBy: string; paginateField: any };
+    after?: { sortBy: string; paginateField: any };
+    sortBy: string;
+    sortDirection: DirectionEnum;
+    paginateField: string;
+    enhanceQuery?: (query: QueryWithHelpers<Array<T_DBModel>, T_DBModel>) => any;
+  }): Promise<{ data: T_MappedEntity[]; next: string | null; previous: string | null }> {
+    if (before && after) {
+      throw new DalException('Cannot specify both "before" and "after" cursors at the same time.');
+    }
+
+    const isDesc = sortDirection === DirectionEnum.DESC;
+    const sortValue = isDesc ? -1 : 1;
+
+    const paginationQuery: any = { ...query };
+
+    if (before) {
+      paginationQuery.$or = [
+        {
+          [sortBy]: isDesc ? { $gt: before.sortBy } : { $lt: before.sortBy },
+        },
+        {
+          $and: [
+            { [sortBy]: { $eq: before.sortBy } },
+            { [paginateField]: isDesc ? { $gt: before.paginateField } : { $lt: before.paginateField } },
+          ],
+        },
+      ];
+    } else if (after) {
+      paginationQuery.$or = [
+        {
+          [sortBy]: isDesc ? { $lt: after.sortBy } : { $gt: after.sortBy },
+        },
+        {
+          $and: [
+            { [sortBy]: { $eq: after.sortBy } },
+            { [paginateField]: isDesc ? { $lt: after.paginateField } : { $gt: after.paginateField } },
+          ],
+        },
+      ];
+    }
+
+    let builder = this.MongooseModel.find(paginationQuery)
+      .sort({ [sortBy]: sortValue, [paginateField]: sortValue })
+      .limit(limit + 1);
+
+    if (enhanceQuery) {
+      builder = enhanceQuery(builder);
+    }
+
+    const rawResults = await builder.exec();
```

Comment:
what is `enhanceQuery`?

now sure i understand the flow here, won't line 383 will be executed at the time as Promise?